src/conversion/onnx_to_krnl/nn/pooling.cpp

```diff
@@ -222,45 +222,107 @@ struct ONNXMaxPoolSingleOutOpLowering : public ConversionPattern {
 
         // 3.1 Prepare indices for accesing the data tensor.
         SmallVector<Value, 4> dataIndices;
-        // Batch indices: n, c
+        // 3.1.1 Batch indices: n, c
         for (int i = 0; i < batchRank; ++i)
           dataIndices.emplace_back(outerLoops.getInductionVar(i));
-        // Spatial indices: sX * rX + kX
+        // 3.1.2 Insert spatial indices: sX * rX + kX
+        // Spatial index may go out of the input index space due to dilation or
+        // ceilMode. Hence, we should keep trace of out-of-index.
+        Value outOfIndex = nullptr;
         for (int i = batchRank; i < nOuterLoops; ++i) {
           Value spatialIndex = outerLoops.getInductionVar(i);
-          // If strides are present then emit the correct access index.
-          if (stridesAttribute && strides[i - batchRank] > 1) {
-            spatialIndex = rewriter.create<MulIOp>(loc,
-                rewriter.create<ConstantIndexOp>(loc, strides[i - batchRank]),
-                outerLoops.getInductionVar(i));
+          // If strides are present (not default) then emit the correct access
+          // index.
+          // sX * rX
+          if (strides[i - batchRank] > 1) {
+            auto strideIndex = emitConstantOp(
+                rewriter, loc, rewriter.getIndexType(), strides[i - batchRank]);
+            spatialIndex = rewriter.create<MulIOp>(
+                loc, strideIndex, outerLoops.getInductionVar(i));
           }
+
+          // sX * rX + kX
           spatialIndex = rewriter.create<AddIOp>(
               loc, spatialIndex, innerLoops.getInductionVar(i - batchRank));
-          // If ceil mode is enabled, then the calculated access index may
-          // exceed its dimension. In such a case, we will use the maximum
-          // index, which causes multiple visits to the element of the
-          // maximum index.
-          // TODO: Avoid multiple visits.
-          if (ceilMode) {
-            Value inputIndex;
+
+          // Dilate the kernel index only if the dilation value is not one (not
```

Comment:
I don't really understand why we have to do this, and maybe it is needed, but I would like to know why. I just thought that dilation, we just start at 0, and then add "dilation" to get to the next filter value. Did you find something online that pushed you toward this implementation?

Reply:
I see. It is perfect to just calculate an offset from the current kernel index. I was doing dilation by expanding the filter window from its center. Now I change to use your method. Thanks!

Reply:
thanks, let me know when its complete